File: sdks/python/aether_sdk/event/schema.py
"""
Schema Registry for Event Validation

Provides schema registry capabilities for event validation, schema evolution,
 and compatibility checking.

Example:
    from aether_sdk.event import SchemaRegistry, InMemorySchemaRegistry, Schema
    
    # Create registry
    registry = InMemorySchemaRegistry()
    
    # Register a schema
    schema = Schema(
        name="UserCreated",
        type="json",
        definition={
            "type": "object",
            "properties": {
                "userId": {"type": "string"},
                "email": {"type": "string"}
            },
            "required": ["userId", "email"]
        }
    )
    await registry.register("UserCreated", schema)
    
    # Validate an event
    event = {"userId": "123", "email": "user@example.com"}
    valid = await registry.validate("UserCreated", event)
"""

from __future__ import annotations
from dataclasses import dataclass, field
from datetime import datetime
from typing import (
    Any,
    Dict,
    List,
    Optional,
    Set,
    Union,
)
from abc import ABC, abstractmethod
from enum import Enum
import asyncio
import json
import uuid

from ..exceptions import AetherError


class Compatibility(Enum):
    """
    Schema compatibility levels for version evolution.
    
    - BACKWARD: New schema can read old data
    - FORWARD: Old schema can read new data  
    - FULL: Both backward and forward compatible
    - NONE: No compatibility (breaking change)
    """
    NONE = "none"
    BACKWARD = "backward"
    FORWARD = "forward"
    FULL = "full"


@dataclass
class SchemaVersion:
    """
    Represents a versioned schema in the registry.
    """
    version: str  # Semantic version like "1.0.0"
    schema_id: str
    definition: Dict[str, Any]
    created_at: datetime = field(default_factory=datetime.utcnow)
    deprecated: bool = False
    compatibility: Compatibility = Compatibility.BACKWARD
    
    def __str__(self) -> str:
        return f"{self.schema_id}@{self.version}"


    
    @property
    def major_version(self) -> int:
        """Extract major version number."""
        return int(self.version.split('.')[0])


@dataclass
class Schema:
    """
    Schema definition for event validation.
    """
    name: str
    type: str  # json, avro, protobuf, custom
    definition: Dict[str, Any]
    description: str = ""
    version: str = "1.0.0"
    namespace: Optional[str] = None
    owner: Optional[str] = None
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)
    
    def to_dict(self) -> Dict[str, Any]:
        """Serialize schema to dictionary."""
        return {
            "name": self.name,
            "type": self.type,
            "definition": self.definition,
            "description": self.description,
            "version": self.version,
            "namespace": self.namespace,
            "owner": self.owner,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> Schema:
        """Deserialize schema from dictionary."""
        return cls(
            name=data["name"],
            type=data["type"],
            definition=data["definition"],
            description=data.get("description", ""),
            version=data.get("version", "1.0.0"),
            namespace=data.get("namespace"),
            owner=data.get("owner"),
            created_at=datetime.fromisoformat(data["created_at"]) if "created_at" in data else datetime.utcnow(),
            updated_at=datetime.fromisoformat(data["updated_at"]) if "updated_at" in data else datetime.utcnow(),
        )


class SchemaError(AetherError):
    """Raised when schema validation fails."""
    
    def __init__(self, message: str, schema_name: Optional[str] = None):
        self.schema_name = schema_name
        super().__init__(message)


class SchemaValidator(ABC):
    """
    Abstract base class for schema validators.
    
    Different validators can be implemented for different
    schema types (JSON Schema, Avro, Protobuf, etc.)
    """
    
    @abstractmethod
    def validate(self, data: Any, schema: Schema) -> List[str]:
        """
        Validate data against a schema.
        
        Args:
            data: The data to validate
            schema: The schema to validate against
        
        Returns:
            List of validation errors (empty if valid)
        """
        pass
    
    @abstractmethod
    def check_compatibility(
        self,
        old_schema: Schema,
        new_schema: Schema
    ) -> Compatibility:
        """
        Check compatibility between two schema versions.
        
        Args:
            old_schema: The current schema
            new_schema: The proposed new schema
        
        Returns:
            Compatibility level between the schemas
        """
        pass


class JsonSchemaValidator(SchemaValidator):
    """
    JSON Schema validator implementation.
    """
    
    def validate(self, data: Any, schema: Schema) -> List[str]:
        """Validate data against JSON Schema."""
        if schema.type != "json":
            return [f"Expected JSON schema, got {schema.type}"]
        
        definition = schema.definition
        errors = []
        
        # Simple JSON Schema validation
        if "required" in definition:
            for field in definition["required"]:
                if field not in data:
                    errors.append(f"Missing required field: {field}")
        
        if "properties" in definition:
            for prop_name, prop_def in definition["properties"].items():
                if prop_name in data:
                    prop_type = prop_def.get("type")
                    if prop_type:
                        if not self._check_type(data[prop_name], prop_type):
                            errors.append(
                                f"Field '{prop_name}' has wrong type: expected {prop_type}"
                            )
        
        return errors
    
    def _check_type(self, value: Any, expected_type: str) -> bool:
        """Check if value matches expected JSON Schema type."""
        type_map = {
            "string": str,
            "number": (int, float),
            "integer": int,
            "boolean": bool,
            "array": list,
            "object": dict,
            "null": type(None),
        }
        
        expected = type_map.get(expected_type)
        if expected is None:
            return True  # Unknown type, skip validation
        
        # Handle number/integer specially
        if expected_type == "number":
            return isinstance(value, (int, float))
        if expected_type == "integer":
            return isinstance(value, int) and not isinstance(value, bool)
        
        return isinstance(value, expected)
    
    def check_compatibility(
        self,
        old_schema: Schema,
        new_schema: Schema
    ) -> Compatibility:
        """Check JSON Schema compatibility."""
        if old_schema.type != new_schema.type:
            return Compatibility.NONE
        
        old_def = old_schema.definition
        new_def = new_schema.definition
        
        # Check required fields - adding required fields is breaking
        old_required = set(old_def.get("required", []))
        new_required = set(new_def.get("required", []))
        
        # New required fields = breaking change
        added_required = new_required - old_required
        if added_required:
            return Compatibility.NONE
        
        # Check property types - changing types is breaking
        old_props = old_def.get("properties", {})
        new_props = new_def.get("properties", {})
        
        for prop_name, old_prop in old_props.items():
            new_prop = new_props.get(prop_name)
            if new_prop and old_prop.get("type") != new_prop.get("type"):
                return Compatibility.NONE
        
        # Adding optional fields = backward compatible
        # Removing optional fields = forward compatible
        # No changes = full compatible
        
        if not new_required and set(new_props.keys()) >= set(old_props.keys()):
            return Compatibility.BACKWARD
        elif new_required or set(new_props.keys()) < set(old_props.keys()):
            return Compatibility.FORWARD
        else:
            return Compatibility.FULL


class SchemaRegistry(ABC):
    """
    Abstract base class for schema registries.
    
    Provides schema registration, validation, and versioning.
    """
    
    @abstractmethod
    async def register(
        self,
        name: str,
        schema: Schema,
        compatibility: Compatibility = Compatibility.BACKWARD
    ) -> SchemaVersion:
        """
        Register a new schema or schema version.
        
        Args:
            name: Schema name
            schema: Schema definition
            compatibility: Expected compatibility level
        
        Returns:
            The created schema version
        """
        pass
    
    @abstractmethod
    async def get_schema(self, name: str, version: Optional[str] = None) -> Optional[Schema]:
        """
        Get a schema by name and optional version.
        
        Args:
            name: Schema name
            version: Optional specific version (latest if None)
        
        Returns:
            The schema or None if not found
        """
        pass
    
    @abstractmethod
    async def get_versions(self, name: str) -> List[SchemaVersion]:
        """
        Get all versions of a schema.
        
        Args:
            name: Schema name
        
        Returns:
            List of schema versions
        """
        pass
    
    @abstractmethod
    async def validate(self, name: str, data: Any, version: Optional[str] = None) -> bool:
        """
        Validate data against a schema.
        
        Args:
            name: Schema name
            data: Data to validate
            version: Optional specific version (latest if None)
        
        Returns:
            True if valid, raises SchemaError if invalid
        """
        pass
    
    @abstractmethod
    async def check_compatibility(
        self,
        name: str,
        old_version: str,
        new_version: str
    ) -> Compatibility:
        """
        Check compatibility between two schema versions.
        
        Args:
            name: Schema name
            old_version: Current version
            new_version: Proposed version
        
        Returns:
            Compatibility level
        """
        pass


class InMemorySchemaRegistry(SchemaRegistry):
    """
    In-memory implementation of SchemaRegistry for testing and development.
    """
    
    def __init__(self, validator: Optional[SchemaValidator] = None):
        self._schemas: Dict[str, List[SchemaVersion]] = {}
        self._validator = validator or JsonSchemaValidator()
        self._lock = asyncio.Lock()
    
    async def register(
        self,
        name: str,
        schema: Schema,
        compatibility: Compatibility = Compatibility.BACKWARD
    ) -> SchemaVersion:
        """Register a new schema or schema version."""
        async with self._lock:
            if name not in self._schemas:
                self._schemas[name] = []
            
            # Determine version number
            existing_versions = self._schemas[name]
            if existing_versions:
                last_version = existing_versions[-1].version
                major, minor, patch = map(int, last_version.split('.'))
                new_version = f"{major}.{minor}.{patch + 1}"
            else:
                new_version = "1.0.0"
            
            # Check compatibility with previous version
            if existing_versions:
                old_schema = Schema(
                    name=name,
                    type=schema.type,
                    definition=existing_versions[-1].definition
                )
                computed_compat = self._validator.check_compatibility(
                    old_schema, schema
                )
                if computed_compat == Compatibility.NONE and compatibility != Compatibility.NONE:
                    # Warning: breaking change detected but allowed
                    pass
            
            schema_version = SchemaVersion(
                version=new_version,
                schema_id=str(uuid.uuid4()),
                definition=schema.definition,
                compatibility=compatibility
            )
            
            self._schemas[name].append(schema_version)
            return schema_version
    
    async def get_schema(self, name: str, version: Optional[str] = None) -> Optional[Schema]:
        """Get a schema by name and optional version."""
        async with self._lock:
            if name not in self._schemas:
                return None
            
            versions = self._schemas[name]
            if not versions:
                return None
            
            if version:
                for sv in versions:
                    if sv.version == version:
                        return Schema(
                            name=name,
                            type="json",  # Assuming JSON for now
                            definition=sv.definition,
                            version=sv.version
                        )
                return None
            
            # Return latest version
            latest = versions[-1]
            return Schema(
                name=name,
                type="json",
                definition=latest.definition,
                version=latest.version
            )
    
    async def get_versions(self, name: str) -> List[SchemaVersion]:
        """Get all versions of a schema."""
        async with self._lock:
            return list(self._schemas.get(name, []))
    
    async def validate(self, name: str, data: Any, version: Optional[str] = None) -> bool:
        """Validate data against a schema."""
        schema = await self.get_schema(name, version)
        if schema is None:
            raise SchemaError(f"Schema not found: {name}", name)
        
        errors = self._validator.validate(data, schema)
        if errors:
            raise SchemaError(
                f"Validation failed: {'; '.join(errors)}",
                name
            )
        
        return True
    
    async def check_compatibility(
        self,
        name: str,
        old_version: str,
        new_version: str
    ) -> Compatibility:
        """Check compatibility between two schema versions."""
        old_schema = await self.get_schema(name, old_version)
        new_schema = await self.get_schema(name, new_version)
        
        if old_schema is None or new_schema is None:
            raise SchemaError(f"Schema version not found: {name}")
        
        return self._validator.check_compatibility(old_schema, new_schema)


__all__ = [
    "Compatibility",
    "SchemaVersion",
    "Schema",
    "SchemaError",
    "SchemaValidator",
    "JsonSchemaValidator",
    "SchemaRegistry",
    "InMemorySchemaRegistry",
]

File: sdks/javascript/src/event/schema.ts
/**
 * Schema Registry for Event Validation
 *
 * Provides schema registration, validation, compatibility checking, and
 * evolution support for event schemas.
 *
 * @example
 * ```typescript
 * import { SchemaRegistry, SchemaCompatibilityMode } from 'aether-sdk/event';
 *
 * const registry = new SchemaRegistry();
 *
 * await registry.register({
 *   name: 'UserCreated',
 *   version: '1.0.0',
 *   type: 'json',
 *   fields: [
 *     { name: 'userId', type: 'string', required: true },
 *     { name: 'email', type: 'string', required: true },
 *   ],
 *   schema: {
 *     type: 'object',
 *     properties: {
 *       userId: { type: 'string' },
 *       email: { type: 'string' },
 *     },
 *     required: ['userId', 'email'],
 *   },
 * });
 *
 * await registry.validate('UserCreated', { userId: '1', email: 'a@b.com' });
 * ```
 *
 * @module aether/event/schema
 */

import {
  SchemaDefinition,
  SchemaField,
  SchemaCompatibilityMode,
} from './types';

/** @internal */
function generateId(): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return crypto.randomUUID();
  }
  return `${Date.now()}-${Math.random().toString(36).slice(2, 11)}`;
}

/**
 * Thrown when schema validation fails.
 */
export class SchemaValidationError extends Error {
  /** Schema name that failed validation. */
  public readonly schemaName: string;
  /** Field-level validation errors. */
  public readonly fieldErrors: FieldValidationError[];

  /**
   * @param message     - Human-readable error summary.
   * @param schemaName  - Name of the schema.
   * @param fieldErrors - Individual field validation failures.
   */
  constructor(
    message: string,
    schemaName: string,
    fieldErrors: FieldValidationError[] = [],
  ) {
    super(message);
    this.name = 'SchemaValidationError';
    this.schemaName = schemaName;
    this.fieldErrors = fieldErrors;
  }
}

/**
 * A single field-level validation error.
 */
export interface FieldValidationError {
  /** Field name that failed validation. */
  field: string;
  /** Human-readable error description. */
  message: string;
  /** Expected type or constraint. */
  expected?: string;
  /** Actual value received. */
  actual?: unknown;
}

/**
 * Schema compatibility check result.
 */
export interface CompatibilityResult {
  /** Whether the schemas are compatible under the requested mode. */
  compatible: boolean;
  /** Detected compatibility level. */
  mode: SchemaCompatibilityMode;
  /** Human-readable description of changes. */
  description: string;
  /** List of specific compatibility issues (empty if compatible). */
  issues: string[];
}

/**
 * Internal versioned schema entry in the registry.
 *
 * @internal
 */
interface VersionedSchema {
  /** Schema ID (UUID). */
  schemaId: string;
  /** Schema definition. */
  schema: SchemaDefinition;
  /** When this version was registered. */
  createdAt: Date;
  /** Whether this version is deprecated. */
  deprecated: boolean;
  /** Compatibility mode enforced at registration time. */
  compatibility: SchemaCompatibilityMode;
}

/**
 * JSON Schema type mapping to JavaScript runtime types.
 *
 * @internal
 */
const JSON_TYPE_MAP: Record<string, (v: unknown) => boolean> = {
  string: (v) => typeof v === 'string',
  number: (v) => typeof v === 'number',
  integer: (v) => typeof v === 'number' && Number.isInteger(v),
  boolean: (v) => typeof v === 'boolean',
  array: (v) => Array.isArray(v),
  object: (v) => typeof v === 'object' && v !== null && !Array.isArray(v),
  null: (v) => v === null,
};

/**
 * Schema registry for managing event schema definitions.
 *
 * Supports registration, validation, compatibility checking, and
 * schema evolution with configurable compatibility modes.
 *
 * @example
 * ```typescript
 * const registry = new SchemaRegistry();
 *
 * // Register initial schema
 * await registry.register({
 *   name: 'OrderCreated',
 *   version: '1.0.0',
 *   type: 'json',
 *   fields: [
 *     { name: 'orderId', type: 'string', required: true },
 *     { name: 'total', type: 'number', required: true },
 *   ],
 *   schema: {
 *     type: 'object',
 *     properties: {
 *       orderId: { type: 'string' },
 *       total: { type: 'number' },
 *     },
 *     required: ['orderId', 'total'],
 *   },
 * });
 *
 * // Validate an event
 * try {
 *   await registry.validate('OrderCreated', { orderId: '123', total: 99.99 });
 * } catch (e) {
 *   if (e instanceof SchemaValidationError) {
 *     console.error(e.fieldErrors);
 *   }
 * }
 *
 * // Evolve the schema
 * const evolved = await registry.evolve(
 *   {
 *     name: 'OrderCreated',
 *     version: '1.1.0',
 *     type: 'json',
 *     fields: [
 *       { name: 'orderId', type: 'string', required: true },
 *       { name: 'total', type: 'number', required: true },
 *       { name: 'currency', type: 'string', required: false, defaultValue: 'USD' },
 *     ],
 *     schema: {
 *       type: 'object',
 *       properties: {
 *         orderId: { type: 'string' },
 *         total: { type: 'number' },
 *         currency: { type: 'string' },
 *       },
 *       required: ['orderId', 'total'],
 *     },
 *   },
 *   '1.1.0',
 *   SchemaCompatibilityMode.BACKWARD,
 * );
 * ```
 */
export class SchemaRegistry {
  private readonly schemas = new Map<string, VersionedSchema[]>();

  /**
   * Register a new schema definition.
   *
   * If a schema with the same name already exists, this registers a new
   * version. The first registration always succeeds. Subsequent registrations
   * are checked for compatibility unless the mode is `NONE`.
   *
   * @param schema - The schema definition to register.
   * @returns The registered schema definition (with assigned version).
   * @throws {SchemaValidationError} If the new schema is incompatible.
   */
  async register(schema: SchemaDefinition): Promise<SchemaDefinition> {
    const name = schema.name;
    let versions = this.schemas.get(name);

    if (!versions) {
      versions = [];
      this.schemas.set(name, versions);
    }

    if (versions.length > 0) {
      const last = versions[versions.length - 1];
      const result = this.checkCompatibility(
        last.schema.schema,
        schema.schema,
        last.schema.fields,
        schema.fields,
      );

      if (!result.compatible && last.compatibility !== SchemaCompatibilityMode.NONE) {
        throw new SchemaValidationError(
          `Schema '${name}' evolution is not compatible: ${result.description}`,
          name,
          result.issues.map((issue) => ({
            field: '',
            message: issue,
          })),
        );
      }
    }

    const versionedSchema: VersionedSchema = {
      schemaId: generateId(),
      schema,
      createdAt: new Date(),
      deprecated: false,
      compatibility: SchemaCompatibilityMode.BACKWARD,
    };

    versions.push(versionedSchema);
    return schema;
  }

  /**
   * Validate data against a registered schema.
   *
   * Performs required-field checks, type checks, and applies default values
   * for missing optional fields.
   *
   * @param eventName - Schema name to validate against.
   * @param data      - Data to validate.
   * @param version   - Optional specific schema version (defaults to latest).
   * @returns `true` if validation passes.
   * @throws {SchemaValidationError} If validation fails or schema not found.
   */
  async validate(
    eventName: string,
    data: unknown,
    version?: string,
  ): Promise<boolean> {
    const schema = this.getSchemaInternal(eventName, version);
    if (!schema) {
      throw new SchemaValidationError(
        `Schema not found: ${eventName}`,
        eventName,
      );
    }

    const errors = this.validateData(data, {
      schemaId: '',
      schema,
      createdAt: new Date(),
      deprecated: false,
      compatibility: SchemaCompatibilityMode.BACKWARD,
    });
    if (errors.length > 0) {
      throw new SchemaValidationError(
        `Validation failed for '${eventName}': ${errors.map((e) => e.message).join('; ')}`,
        eventName,
        errors,
      );
    }

    return true;
  }

  /**
   * Get a schema definition by name and optional version.
   *
   * @param eventName - Schema name.
   * @param version   - Optional version string (latest if omitted).
   * @returns The schema definition, or `undefined` if not found.
   */
  async getSchema(
    eventName: string,
    version?: string,
  ): Promise<SchemaDefinition | undefined> {
    return this.getSchemaInternal(eventName, version);
  }

  /**
   * Evolve a schema to a new version with compatibility enforcement.
   *
   * Validates that the new schema is compatible with the latest version
   * under the specified compatibility mode, then registers it.
   *
   * @param schema             - New schema definition.
   * @param newVersion         - Target version string.
   * @param compatibilityMode  - Compatibility mode to enforce.
   * @returns The evolved schema definition.
   * @throws {SchemaValidationError} If the schemas are not compatible.
   */
  async evolve(
    schema: SchemaDefinition,
    newVersion: string,
    compatibilityMode: SchemaCompatibilityMode,
  ): Promise<SchemaDefinition> {
    const name = schema.name;
    const versions = this.schemas.get(name);

    if (!versions || versions.length === 0) {
      return this.register(schema);
    }

    const last = versions[versions.length - 1];
    const result = this.checkCompatibility(
      last.schema.schema,
      schema.schema,
      last.schema.fields,
      schema.fields,
    );

    if (compatibilityMode === SchemaCompatibilityMode.FULL) {
      if (result.mode !== SchemaCompatibilityMode.FULL) {
        throw new SchemaValidationError(
          `Schema '${name}' is not fully compatible: ${result.description}`,
          name,
          result.issues.map((issue) => ({ field: '', message: issue })),
        );
      }
    } else if (compatibilityMode === SchemaCompatibilityMode.BACKWARD) {
      if (
        result.mode !== SchemaCompatibilityMode.BACKWARD &&
        result.mode !== SchemaCompatibilityMode.FULL
      ) {
        throw new SchemaValidationError(
          `Schema '${name}' is not backward compatible: ${result.description}`,
          name,
          result.issues.map((issue) => ({ field: '', message: issue })),
        );
      }
    } else if (compatibilityMode === SchemaCompatibilityMode.FORWARD) {
      if (
        result.mode !== SchemaCompatibilityMode.FORWARD &&
        result.mode !== SchemaCompatibilityMode.FULL
      ) {
        throw new SchemaValidationError(
          `Schema '${name}' is not forward compatible: ${result.description}`,
          name,
          result.issues.map((issue) => ({ field: '', message: issue })),
        );
      }
    }

    const evolved: SchemaDefinition = {
      ...schema,
      version: newVersion,
    };

    const versionedSchema: VersionedSchema = {
      schemaId: generateId(),
      schema: evolved,
      createdAt: new Date(),
      deprecated: false,
      compatibility: compatibilityMode,
    };

    versions.push(versionedSchema);
    return evolved;
  }

  /**
   * Get all registered versions for a schema.
   *
   * @param eventName - Schema name.
   * @returns Array of version strings.
   */
  async getVersions(eventName: string): Promise<string[]> {
    const versions = this.schemas.get(eventName);
    if (!versions) {
      return [];
    }
    return versions.map((v) => v.schema.version);
  }

  /**
   * Check compatibility between two JSON Schema definitions.
   *
   * @param oldSchema    - Old schema definition object.
   * @param newSchema    - New schema definition object.
   * @param oldFields    - Old field definitions.
   * @param newFields    - New field definitions.
   * @returns Compatibility check result.
   */
  private checkCompatibility(
    oldSchema: Record<string, unknown>,
    newSchema: Record<string, unknown>,
    oldFields: SchemaField[],
    newFields: SchemaField[],
  ): CompatibilityResult {
    const issues: string[] = [];

    const oldDef = oldSchema as Record<string, unknown>;
    const newDef = newSchema as Record<string, unknown>;

    const oldProps = (oldDef.properties as Record<string, unknown>) ?? {};
    const newProps = (newDef.properties as Record<string, unknown>) ?? {};

    const oldRequired = new Set(
      (oldDef.required as string[]) ?? [],
    );
    const newRequired = new Set(
      (newDef.required as string[]) ?? [],
    );

    const oldFieldMap = new Map(oldFields.map((f) => [f.name, f]));
    const newFieldMap = new Map(newFields.map((f) => [f.name, f]));

    const oldPropNames = new Set(Object.keys(oldProps));
    const newPropNames = new Set(Object.keys(newProps));

    const addedRequired = [...newRequired].filter((f) => !oldRequired.has(f));
    if (addedRequired.length > 0) {
      issues.push(`Added required fields: ${addedRequired.join(', ')}`);
    }

    for (const [name, oldProp] of Object.entries(oldProps)) {
      const newProp = newProps[name];
      if (newProp) {
        const oldType = (oldProp as Record<string, unknown>).type;
        const newType = (newProp as Record<string, unknown>).type;
        if (oldType !== newType) {
          issues.push(`Field '${name}' type changed from ${oldType} to ${newType}`);
        }
      }
    }

    const removedFields = [...oldPropNames].filter((f) => !newPropNames.has(f));
    const addedFields = [...newPropNames].filter((f) => !oldPropNames.has(f));

    const hasTypeChanges = issues.some((i) => i.includes('type changed'));
    const hasBreakingChanges = addedRequired.length > 0 || hasTypeChanges;

    if (hasBreakingChanges) {
      return {
        compatible: false,
        mode: SchemaCompatibilityMode.NONE,
        description: 'Breaking changes detected',
        issues,
      };
    }

    const isSubset = removedFields.length > 0 && addedFields.length === 0;
    const isSuperset = addedFields.length > 0 && removedFields.length === 0;

    if (isSuperset) {
      return {
        compatible: true,
        mode: SchemaCompatibilityMode.BACKWARD,
        description: 'New schema is a superset (backward compatible)',
        issues,
      };
    }

    if (isSubset) {
      return {
        compatible: true,
        mode: SchemaCompatibilityMode.FORWARD,
        description: 'New schema is a subset (forward compatible)',
        issues,
      };
    }

    if (addedFields.length === 0 && removedFields.length === 0) {
      return {
        compatible: true,
        mode: SchemaCompatibilityMode.FULL,
        description: 'No field changes (fully compatible)',
        issues,
      };
    }

    return {
      compatible: false,
      mode: SchemaCompatibilityMode.NONE,
      description: 'Mixed field additions and removals',
      issues: [...issues, 'Both fields added and removed'],
    };
  }

  /**
   * Validate data against a schema definition.
   *
   * @param data   - Data to validate.
   * @param schema - Schema definition to validate against.
   * @returns Array of field validation errors (empty if valid).
   */
  private validateData(
    data: unknown,
    schema: VersionedSchema,
  ): FieldValidationError[] {
    const errors: FieldValidationError[] = [];

    if (typeof data !== 'object' || data === null || Array.isArray(data)) {
      errors.push({
        field: '',
        message: 'Expected object, received non-object value',
        expected: 'object',
        actual: data,
      });
      return errors;
    }

    const record = data as Record<string, unknown>;
    const schemaDef = schema.schema.schema as unknown as Record<string, unknown>;
    const properties = (schemaDef.properties as Record<string, unknown>) ?? {};
    const requiredFields = new Set(
      (schemaDef.required as string[]) ?? [],
    );

    for (const field of schema.schema.fields) {
      if (field.required && !(field.name in record)) {
        if (field.defaultValue !== undefined) {
          record[field.name] = field.defaultValue;
        } else {
          errors.push({
            field: field.name,
            message: `Missing required field '${field.name}'`,
            expected: field.type,
            actual: undefined,
          });
        }
      }
    }

    for (const required of requiredFields) {
      if (!(required in record)) {
        const fieldDef = schema.schema.fields.find((f) => f.name === required);
        if (fieldDef?.defaultValue !== undefined) {
          record[required] = fieldDef.defaultValue;
        } else {
          errors.push({
            field: required,
            message: `Missing required field '${required}'`,
            expected: 'present',
            actual: undefined,
          });
        }
      }
    }

    for (const [propName, propDef] of Object.entries(properties)) {
      if (!(propName in record)) {
        continue;
      }

      const propRecord = propDef as Record<string, unknown>;
      const expectedType = propRecord.type as string | undefined;

      if (expectedType) {
        const checker = JSON_TYPE_MAP[expectedType];
        if (checker && !checker(record[propName])) {
          errors.push({
            field: propName,
            message: `Field '${propName}' has wrong type: expected ${expectedType}`,
            expected: expectedType,
            actual: typeof record[propName],
          });
        }
      }
    }

    return errors;
  }

  /**
   * Internal helper to retrieve a schema by name and version.
   *
   * @param eventName - Schema name.
   * @param version   - Optional version string.
   * @returns Schema definition or `undefined`.
   */
  private getSchemaInternal(
    eventName: string,
    version?: string,
  ): SchemaDefinition | undefined {
    const versions = this.schemas.get(eventName);
    if (!versions || versions.length === 0) {
      return undefined;
    }

    if (version) {
      const match = versions.find((v) => v.schema.version === version);
      return match?.schema;
    }

    return versions[versions.length - 1].schema;
  }
}
